core/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -527,172 +544,49 @@ static ReplicaItemExecutionMode replicaItemExecutionMode(final BulkItemRequest r
-                SourceToParse.source(shardId.getIndexName(),
-                        request.type(), request.id(), request.source(), request.getContentType())
-                .routing(request.routing()).parent(request.parent());
-        final VersionType versionType = request.versionType().versionTypeForReplicationAndRecovery();
-        assert versionType.validateVersionForWrites(version);
-
-        return replica.prepareIndexOnReplica(sourceToParse, seqNo, primaryTerm, version, versionType,
-                request.getAutoGeneratedTimestamp(), request.isRetry());
-    }
-
-    /** Utility method to prepare an index operation on primary shards */
-    private static Engine.Index prepareIndexOperationOnPrimary(IndexRequest request, IndexShard primary) {
-        final SourceToParse sourceToParse =
-                SourceToParse.source(request.index(), request.type(),
-                        request.id(), request.source(), request.getContentType())
-                .routing(request.routing()).parent(request.parent());
-        return primary.prepareIndexOnPrimary(sourceToParse, request.version(), request.versionType(),
-                request.getAutoGeneratedTimestamp(), request.isRetry());
-    }
-
     /** Executes index operation on primary shard after updates mapping if dynamic mappings are found */
     static Engine.IndexResult executeIndexRequestOnPrimary(IndexRequest request, IndexShard primary,
                                                            MappingUpdatePerformer mappingUpdater) throws Exception {
-        // Update the mappings if parsing the documents includes new dynamic updates
-        final Engine.Index preUpdateOperation;
-        final Mapping mappingUpdate;
-        final boolean mappingUpdateNeeded;
+        final SourceToParse sourceToParse =
+            SourceToParse.source(request.index(), request.type(), request.id(), request.source(), request.getContentType())
+                .routing(request.routing()).parent(request.parent());
         try {
-            preUpdateOperation = prepareIndexOperationOnPrimary(request, primary);
-            mappingUpdate = preUpdateOperation.parsedDoc().dynamicMappingsUpdate();
-            mappingUpdateNeeded = mappingUpdate != null;
-            if (mappingUpdateNeeded) {
-                mappingUpdater.updateMappings(mappingUpdate, primary.shardId(), request.type());
-            }
-        } catch (MapperParsingException | IllegalArgumentException failure) {
-            return new Engine.IndexResult(failure, request.version());
+            return primary.applyIndexOperationOnPrimary(request.version(), request.versionType(), sourceToParse,
+                request.getAutoGeneratedTimestamp(), request.isRetry(), update -> {
+                    mappingUpdater.updateMappings(update, primary.shardId(), sourceToParse.type());
+                    throw new ReplicationOperation.RetryOnPrimaryException(primary.shardId(), "Mapping updated");
```

Comment:
I've added a comment and a TODO how we could possibly simplify this. I think this needs a bit more study though, and leave that to a future PR.